tokio-net/src/driver/sharded_slab/tests.rs

```diff
@@ -0,0 +1,222 @@
+        assert_eq!(s.get(idx1), Some(&5));
+        let idx2 = s.insert(6).expect("insert");
+        assert_eq!(s.get(idx2), Some(&6));
+        assert_eq!(s.remove(idx1), Some(5));
+        assert_eq!(s.get(idx1), None);
+        assert_eq!(s.get(idx2), Some(&6));
+        assert_eq!(s.remove(idx2), Some(6));
+        assert_eq!(s.get(idx2), None);
+
+        t1.join().expect("thread 1 should not panic");
+        t2.join().expect("thread 2 should not panic");
+    });
+}
+
+#[test]
+fn remove_remote() {
+    loom::model(|| {
+        let slab = Arc::new(Slab::new());
+
+        let idx1 = slab.insert(1).expect("insert");
+        assert_eq!(slab.get(idx1), Some(&1));
+
+        let idx2 = slab.insert(2).expect("insert");
+        assert_eq!(slab.get(idx2), Some(&2));
+
+        let idx3 = slab.insert(3).expect("insert");
+        assert_eq!(slab.get(idx3), Some(&3));
+
+        let s = slab.clone();
+        let t1 = thread::spawn(move || {
+            assert_eq!(s.get(idx2), Some(&2));
+            assert_eq!(s.remove(idx2), Some(2));
+        });
+
+        let s = slab.clone();
+        let t2 = thread::spawn(move || {
+            assert_eq!(s.get(idx3), Some(&3));
+            assert_eq!(s.remove(idx3), Some(3));
+        });
+
+        t1.join().expect("thread 1 should not panic");
+        t2.join().expect("thread 2 should not panic");
+
+        assert_eq!(slab.get(idx1), Some(&1));
+        assert_eq!(slab.get(idx2), None);
+        assert_eq!(slab.get(idx3), None);
+    });
+}
+
+#[test]
+fn concurrent_insert_remove() {
+    loom::model(|| {
+        let slab = Arc::new(Slab::new());
+        let pair = Arc::new((Mutex::new(None), Condvar::new()));
+
+        let slab2 = slab.clone();
+        let pair2 = pair.clone();
+        let remover = thread::spawn(move || {
+            let (lock, cvar) = &*pair2;
+            for i in 0..2 {
+                println!("--- remover i={} ---", i);
+                let mut next = lock.lock().unwrap();
+                while next.is_none() {
+                    next = cvar.wait(next).unwrap();
+                }
+                let key = next.take().unwrap();
+                assert_eq!(slab2.remove(key), Some(i));
+                cvar.notify_one();
+            }
+        });
+
+        let (lock, cvar) = &*pair;
+        for i in 0..2 {
+            println!("--- inserter i={} ---", i);
+            let key = slab.insert(i).expect("insert");
+
+            let mut next = lock.lock().unwrap();
+            *next = Some(key);
+            cvar.notify_one();
+
+            // Wait for the item to be removed.
+            while next.is_some() {
+                next = cvar.wait(next).unwrap();
+            }
+
+            assert!(slab.get(key).is_none());
+        }
+
+        remover.join().unwrap();
+    })
+}
+
+// We can no longer easily implement this test easily, since we can't override the
+// slab's page size, and thus we would have to put 32 elements into the slab,
```

Comment:
It *should* be possible to alter the slab's page size for loom tests. I've done this in `mpsc` (for the same reasons).

https://github.com/tokio-rs/tokio/blob/master/tokio-sync/tests/fuzz_list.rs#L20